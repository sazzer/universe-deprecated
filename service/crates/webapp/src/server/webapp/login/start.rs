use crate::server::request_id::RequestId;
use rocket::get;
use universe_templates::Template;

#[get("/login")]
#[tracing::instrument]
pub fn start_login(_request_id: RequestId) -> Template {
    Template::new("login/start.tera")
}
