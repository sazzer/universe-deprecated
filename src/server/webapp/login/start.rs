use crate::server::request_id::RequestId;
use crate::server::webapp::templates::Template;
use rocket::get;

#[get("/login")]
#[tracing::instrument]
pub fn start_login(_request_id: RequestId) -> Template {
    Template::new("login/start.tera")
}
