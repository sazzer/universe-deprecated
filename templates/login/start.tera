{% extends "layout/landing.tera" %}

{% block sidebar %}
<h3>{{t(key="login_title_login")}}</h3>
<form action="/login" method="POST">
  <div class="form-group">
    <label for="login_username">{{t(key="login_username_label")}}</label>
    <input type="text" class="form-control" id="login_username" autofocus />
  </div>
  <button type="submit" class="btn btn-primary">{{t(key="login_submit_login")}}</button>
</form>
{% endblock sidebar %}
