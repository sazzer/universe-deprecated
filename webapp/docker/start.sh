#!/bin/sh

echo '=== Starting Universe UI ==='
cat /usr/share/nginx/html/config.js.template | envsubst > /usr/share/nginx/html/config.js

echo Configuration:
cat /usr/share/nginx/html/config.js

nginx -g 'daemon off;'
