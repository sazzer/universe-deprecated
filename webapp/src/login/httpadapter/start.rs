use crate::users::{UserService, Username};
use universe_templates::Template;

/// Handle when the initial login form was submitted and we need to look up a username to see if we
/// are performing a Login or a Register action
pub fn start_login(username: Username, user_service: &dyn UserService) -> Template {
    let template = match user_service.username_exists(username.clone()) {
        false => "login/register.tera",
        true => "login/login.tera",
    };

    Template::new(template).with_data("username", username.as_ref())
}
