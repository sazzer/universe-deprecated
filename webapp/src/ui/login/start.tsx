import React, { useState } from 'react';
import { useTranslation } from "react-i18next";
import { useForm, ErrorMessage, FieldValues } from 'react-hook-form';
import * as yup from "yup";

/** Shape of the properties required for the Start Login Form view */
export interface StartLoginFormProps {
  onSubmit: (username: string) => Promise<void>,
}

/**
 * Render the view for the Start Login Form
 */
export const StartLoginForm: React.FC<StartLoginFormProps> = ({ onSubmit }) => {
  const { t } = useTranslation();
  const [pending, setPending] = useState(false);
  const [error, setError] = useState('');

  const { register, errors, handleSubmit } = useForm({
    validationSchema: yup.object().shape({
      username: yup.string()
        .required(t('login.username.errors.required'))
        .trim(),
    }),
    validateCriteriaMode: 'all',
    defaultValues: {
      username: ''
    }
  });

  const onSubmitHandler = async (data: FieldValues) => {
    setPending(true);
    setError('');
    try {
      await onSubmit(data.username.trim());
    } catch (e) {
      setPending(false);
      setError(e.toString());
    }
  };

  let errorMessage;
  if (error) {
    errorMessage = (
      <div className="form-group">
        <div className="alert alert-danger" role="alert">
          {t('errors.unexpected', {
            message: error,
          })}
        </div>
      </div>
    );
  }
  return (
    <>
      <h3>{t('login.start.title')}</h3>
      <form onSubmit={handleSubmit(onSubmitHandler)} data-test="startLoginForm">
        <div className="form-group" data-test="username">
          <label htmlFor="username">{t('login.username.label')}</label>
          <input type="text"
            className={errors.username ? 'form-control is-invalid' : 'form-control'}
            id="username"
            name="username"
            autoFocus
            ref={register} />
          <ErrorMessage errors={errors} name="username" as={<div className="invalid-feedback" />} />
        </div>
        <div className="form-group">
          <button type="submit"
            className="btn btn-primary"
            disabled={pending}>
            {pending && <span className="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>}
            {t('login.start.submit')}
          </button>
        </div>
        {errorMessage}
      </form>
    </>
  );
};
