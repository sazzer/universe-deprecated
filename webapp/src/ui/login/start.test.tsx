import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import { StartLoginForm } from './start';

describe('The Start Login View', () => {
  test('Rendering when not pending', () => {
    const { container } = render(<StartLoginForm onSubmit={async () => { }} />);
    expect(container).toMatchSnapshot();
  });

  test('Entering a username', () => {
    const { container, getByLabelText } = render(<StartLoginForm onSubmit={async () => { }} />);

    fireEvent.change(getByLabelText('Username'), { target: { value: 'testuser' } });
    expect(container).toMatchSnapshot();
  });

  test('Entering a whitespace-only username', () => {
    const { container, getByLabelText } = render(<StartLoginForm onSubmit={async () => { }} />);

    fireEvent.change(getByLabelText('Username'), { target: { value: '  ' } });
    expect(container).toMatchSnapshot();
  });

  describe('Submitting a username', () => {

    test('Successfully', async () => {
      const onSubmit = jest.fn()
      const { container, getByLabelText, getByText } = render(<StartLoginForm onSubmit={onSubmit} />);

      fireEvent.change(getByLabelText('Username'), { target: { value: 'testuser' } });
      fireEvent.click(getByText('Login / Register', { selector: 'button' }));

      await wait(() => expect(onSubmit).toHaveBeenCalledTimes(1));
      expect(onSubmit).toHaveBeenCalledWith('testuser');
      expect(container).toMatchSnapshot();
    });

    test('With a network error', async () => {
      const onSubmit = jest.fn(async () => {
        throw new Error('Network Error');
      });
      const { container, getByLabelText, getByText } = render(<StartLoginForm onSubmit={onSubmit} />);

      fireEvent.change(getByLabelText('Username'), { target: { value: 'testuser' } });
      fireEvent.click(getByText('Login / Register', { selector: 'button' }));

      await wait(() => expect(onSubmit).toHaveBeenCalledTimes(1));
      expect(onSubmit).toHaveBeenCalledWith('testuser');
      expect(container).toMatchSnapshot();
    });

    test('Submitting a blank username', async () => {
      const onSubmit = jest.fn()
      const { container, getByText } = render(<StartLoginForm onSubmit={onSubmit} />);

      fireEvent.click(getByText('Login / Register', { selector: 'button' }));

      await wait(() => expect(onSubmit).toHaveBeenCalledTimes(0));
      expect(container).toMatchSnapshot();
    });

    test('Submitting a whitespace-only username', async () => {
      const onSubmit = jest.fn()
      const { container, getByLabelText, getByText } = render(<StartLoginForm onSubmit={onSubmit} />);

      fireEvent.change(getByLabelText('Username'), { target: { value: '  ' } });
      fireEvent.click(getByText('Login / Register', { selector: 'button' }));

      await wait(() => expect(onSubmit).toHaveBeenCalledTimes(0));
      expect(container).toMatchSnapshot();
    });
  });

});
