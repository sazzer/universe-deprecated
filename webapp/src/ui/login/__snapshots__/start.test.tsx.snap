// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Start Login View Entering a username 1`] = `
<div>
  <h3>
    Login / Register
  </h3>
  <form
    data-test="startLoginForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login / Register
      </button>
    </div>
  </form>
</div>
`;

exports[`The Start Login View Entering a whitespace-only username 1`] = `
<div>
  <h3>
    Login / Register
  </h3>
  <form
    data-test="startLoginForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login / Register
      </button>
    </div>
  </form>
</div>
`;

exports[`The Start Login View Rendering when not pending 1`] = `
<div>
  <h3>
    Login / Register
  </h3>
  <form
    data-test="startLoginForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login / Register
      </button>
    </div>
  </form>
</div>
`;

exports[`The Start Login View Submitting a username Submitting a blank username 1`] = `
<div>
  <h3>
    Login / Register
  </h3>
  <form
    data-test="startLoginForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control is-invalid"
        id="username"
        name="username"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a username
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login / Register
      </button>
    </div>
  </form>
</div>
`;

exports[`The Start Login View Submitting a username Submitting a whitespace-only username 1`] = `
<div>
  <h3>
    Login / Register
  </h3>
  <form
    data-test="startLoginForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control is-invalid"
        id="username"
        name="username"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a username
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login / Register
      </button>
    </div>
  </form>
</div>
`;

exports[`The Start Login View Submitting a username Successfully 1`] = `
<div>
  <h3>
    Login / Register
  </h3>
  <form
    data-test="startLoginForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        <span
          aria-hidden="true"
          class="spinner-border spinner-border-sm"
          role="status"
        />
        Login / Register
      </button>
    </div>
  </form>
</div>
`;

exports[`The Start Login View Submitting a username With a network error 1`] = `
<div>
  <h3>
    Login / Register
  </h3>
  <form
    data-test="startLoginForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login / Register
      </button>
    </div>
    <div
      class="form-group"
    >
      <div
        class="alert alert-danger"
        role="alert"
      >
        An unexpected error occurred: Error: Network Error
      </div>
    </div>
  </form>
</div>
`;
